2_process.smk

```diff
@@ -0,0 +1,155 @@
+    # needed because lake_metadata.csv is used to determine dynamic files
+    lake_metadata_file = checkpoints.fetch_mntoha_metadata.get(**wildcards).output[0]
+    lake_metadata = pd.read_csv(lake_metadata_file)
+    lake = lake_metadata.loc[lake_metadata['site_id']==wildcards.site_id].iloc[0]
+    # also make this function depend on unzip_archive
+    # needed to link unzipped files with unzip_archive rule
+    file_category = 'dynamic_mntoha'
+    drivers_directory = f'inputs_{lake.group_id}'
+    unzip_archive_drivers = checkpoints.unzip_archive.get(
+        file_category=file_category,
+        directory_name=drivers_directory
+    ).output[0]
+    clarity_directory = f'clarity_{lake.group_id}'
+    unzip_archive_clarity = checkpoints.unzip_archive.get(
+        file_category=file_category,
+        directory_name=clarity_directory
+    ).output[0]
+    ice_flags_directory = f'ice_flags_{lake.group_id}'
+    unzip_archive_ice_flags = checkpoints.unzip_archive.get(
+        file_category=file_category,
+        directory_name=ice_flags_directory
+    ).output[0]
+
+    # dynamic filenames
+    dynamic_dir = '1_fetch/out'
+    drivers_file = f'{dynamic_dir}/dynamic_mntoha/inputs_{lake.group_id}/{lake.meteo_filename}'
+    clarity_file = f'{dynamic_dir}/dynamic_mntoha/clarity_{lake.group_id}/gam_{lake.site_id}_clarity.csv'
+    ice_flags_file = f'{dynamic_dir}/dynamic_mntoha/ice_flags_{lake.group_id}/pb0_{lake.site_id}_ice_flags.csv'
+    return [drivers_file, clarity_file, ice_flags_file]
+
+
+# Summarize training sequences
+rule process_mntoha:
+    input:
+        mntoha_lake_sequence_files
+    output:
+        "2_process/out/mntoha_sequences/mntoha_sequences_summary.csv"
+    run:
+        save_sequences_summary(input, output[0])
+
+
+# Create .npy of input/output sequences for one lake to use for training and testing
+rule mntoha_lake_sequences:
+    input:
+        "2_process/tmp/mntoha/lake_metadata_augmented.csv",
+        "2_process/tmp/mntoha/temperature_observations_interpolated.csv",
+        dynamic_filenames
+    output:
+        "2_process/out/mntoha_sequences/sequences_{site_id}.npy"
+    params:
+        config = config
+    script:
+        "2_process/src/lake_sequences_mntoha.py"
+
+
+# Add elevation to MNTOHA lake metadata
+rule augment_mntoha_lake_metadata:
+    input:
+        "1_fetch/out/lake_metadata.csv"
+    output:
+        "2_process/tmp/mntoha/lake_metadata_augmented.csv"
+    script:
+        "2_process/src/make_lake_metadata_augmented.py"
+
+
+def mntoha_obs_file(wildcards):
+    '''
+    Return MNTOHA temperature observations filepath.
+
+    Depend on unzip_archive checkpoint to ensure that
+    temperature_observations.csv gets unzipped.
+
+    :param wildcards: Snakemake wildcards (empty).
+    :returns: Path of MNTOHA temperature observations csv
+
+    '''
+    # Trigger checkpoint to unzip temperature_observations.csv
+    obs_file_directory = checkpoints.unzip_archive.get(
```

Comment:
This is pretty nifty!

Reply:
Just so I'm clear -- the way this system works is that `1_fetch/tmp/obs_mntoha/temperature_observations.zip` doesn't get unzipped until the checkpoint is triggered here?

Reply:
Yes! The pattern here is a simpler version of the pattern involving `dynamic_filenames` and `mntoha_lake_sequences`. It allows dynamic branching. This is the usage pattern for checkpoints that took me some time to wrap my head around!

1. A rule's input is a function.
2. The function triggers a checkpoint. Snakemake knows now that the rule depends on execution of the checkpoint.
3. The checkpoint creates some files.
4. After the checkpoint is run, Snakemake recreates the pipeline graph and reruns the function.
5. The output of the function are the files that are inputs to the rule. Those files will probably include at least some of the files created by the checkpoint.
6. Snakemake's newly recreated graph can include newly created files.
7. Now that its input files have been created, the rule can run.